src/utils/triton.h

```diff
@@ -0,0 +1,180 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#ifndef DALI_BACKEND_UTILS_TRITON_H_
+#define DALI_BACKEND_UTILS_TRITON_H_
+
+#include "src/dali_executor/io_descriptor.h"
+#include "src/dali_executor/utils/dali.h"
+#include "triton/backend/backend_model.h"
+#include "triton/backend/backend_model_instance.h"
+
+namespace triton { namespace backend { namespace dali {
+
+
+/**
+ * @brief Converts TRITONSERVER_DataType to dali_data_type_t
+ */
+inline dali_data_type_t to_dali(TRITONSERVER_DataType t) {
+  assert(t >= 0 && t <= 13);
+  // Use the trick, that TRITONSERVER_DataType and dali_data_type_t
+  // types have more-or-less the same order, with few exceptions
+  if (t == 0)
+    return static_cast<dali_data_type_t>(-1);
+  if (t == 1)
+    return static_cast<dali_data_type_t>(11);
+  if (t == 13)
+    return static_cast<dali_data_type_t>(0);
+  return static_cast<dali_data_type_t>(t - 2);
+}
+
+/**
+ * @brief Converts TRITONSERVER_MemoryType to DALI device_type_t
+ */
+inline device_type_t to_dali(TRITONSERVER_MemoryType t) {
+  switch (t) {
+    case TRITONSERVER_MEMORY_CPU:
+    case TRITONSERVER_MEMORY_CPU_PINNED:
+      return device_type_t::CPU;
+    case TRITONSERVER_MEMORY_GPU:
+      return device_type_t::GPU;
+    default:
+      throw std::invalid_argument("Unknown memory type");
+  }
+}
+
+/**
+ * @brief Converts dali_data_type_t to TRITONSERVER_DataType
+ */
+inline TRITONSERVER_DataType to_triton(dali_data_type_t t) {
+  assert(-1 <= t && t <= 11);
+  // Use the trick, that TRITONSERVER_DataType and dali_data_type_t
+  // types have more-or-less the same order, with few exceptions
+  if (t == -1)
+    return static_cast<TRITONSERVER_DataType>(0);
+  if (t == 11)
+    return static_cast<TRITONSERVER_DataType>(1);
+  return static_cast<TRITONSERVER_DataType>(t + 2);
+}
+
+/**
+ * @brief Converts device_type_t to TRITONSERVER_MemoryType
+ */
+inline TRITONSERVER_MemoryType to_triton(device_type_t dev) {
+  switch (dev) {
+    case CPU:
+      return TRITONSERVER_MEMORY_CPU_PINNED;
+    case GPU:
+      return TRITONSERVER_MEMORY_GPU;
+    default:
+      throw std::invalid_argument("Unknown memory type");
+  }
+}
+
+class TritonInput {
+ public:
+  TritonInput(TRITONBACKEND_Input *handle) : handle_(handle) {
+    TRITONSERVER_DataType input_datatype;
```

Comment:
If we go the class way (I like it btw), maybe `TRITONBACKEND_Input` should be enclosed inside `TritonInput`? Why would we need it outside?

Reply:
I think creating the `TritonInput` object is a responsibility of `TritonRequest` class.